### DISCUSSION TITLE
PINNs for Finance: Results do not converge to expected solution and IC is not obeyed (not even as hard constraint)

### DISCUSSION THREAD

**Discussion by PhiKun24** [lululxvi/deepxde#1710] [General]
Hello Dr. @lululxvi and deepxde community,

I tried to solve a finance PDE with DeepXDE. Unfortunately, I seem to have run into some problems, as I cannot get the model to produce the expected output. The PDE to be solved is 

![BlackScholesPDE2](https://github.com/lululxvi/deepxde/assets/156695486/3c7c541b-6524-496f-85a7-08c0190219c5)

with 

![BlackScholesBoundaryConditions](https://github.com/lululxvi/deepxde/assets/156695486/3ba0ebb4-6feb-44dc-8ace-bbc329f80469)

where the first two are the BCs and the last one is the IC (although in finance its actually the terminal condition).

The expected solution should look sth like this: 

![BlackScholesPDESol1](https://github.com/lululxvi/deepxde/assets/156695486/c18a59b0-9a72-43d9-8e22-95fd1f9991ec)
![BlackScholesPDESol2](https://github.com/lululxvi/deepxde/assets/156695486/8101b255-ff9a-4a0e-b9a4-18a404c4ca1e)

My code and the respective output can be found in the attached notebook. I tried to encode the BC and IC both as soft and hard constraints and also tried different network architectures, parametrisations etc. However, non of it worked and the model does not converge to the expected solution shown above. What's quite surprising to me (as someone who's just getting started with ML) is that i) not even the hard-coded IC is met and ii) more sophisticated networks (i.e. more nodes and/or more layers) as well as more iterations do rather lead to worse results. 

I am sure the problem is sitting in front of my computer, however, I could not find the errors yet. Would be awesome and of great help to get some feedback from the community!

Many thanks in advance :)

[DeepXDE_PINN_OP_v0.2.zip](https://github.com/lululxvi/deepxde/files/14968461/DeepXDE_PINN_OP_v0.2.zip)

**Comment by vl-dud** [comment]
Your `S_max` equals 50. Have you tried working with spatial boundaries from 0 to 1?

**Comment by PhiKun24** [comment]
Many thanks for your suggestion. However, I had tried that already and it did, not significantly change the outcome...

**Comment by vl-dud** [comment]
Your solution function `func_bsm` takes an array of shape (N, 2) and returns an array of shape (2,). Is this correct?
```
def func_bsm(x):
    S, t = x[0], x[1]
    d1 = (np.log(S/K) + (r + vola**2/2)*(T-t)) / (vola*np.sqrt(T-t))
    d2 = d1 - vola * np.sqrt(T-t)
    sol = S * N(d1) - K * np.exp(-r*(T-t))* N(d2)
    return sol
```

**Reply by PhiKun24** [reply]
Thanks again for your support!

I did adjust the function accordingly, but ran into a new problem: "IC function should return an array of shape N by 1 for each component" (see attached nb). I checked the shape of the IC function's output and it should be fine (i.e. Nx1). But when used in model.train() it blows up to (NxN). I'm lacking some intuition for this 'behavior'. Do you have an idea what's going wrong/how to fix it?

[DeepXDE_PINN_OP_workingVersion.zip](https://github.com/lululxvi/deepxde/files/15052015/DeepXDE_PINN_OP_workingVersion.zip)

**Reply by vl-dud** [reply]
Replace `ic_test = np.maximum(0, x-np.full(len(x), K))` with `ic_test = np.maximum(0, x - K)`

**Reply by PhiKun24** [reply]
This works, i.e. the code runs without errors, many thanks! But now we're back at the starting point in the sense that the output still doesnt make any sense. 

Using soft constraints I get sth like 
![results, iter=10000 (2)](https://github.com/lululxvi/deepxde/assets/156695486/eecccedf-bdc6-4bbd-a361-ff14a0878876)

with hard constraints results get even worse...
![results with hard constraints](https://github.com/lululxvi/deepxde/assets/156695486/a1dd11aa-c8cc-4ba9-a094-64901144b849)
